File: typescipt/src/parser/parserContext.ts


export class ParserContext extends IParserContext {
   private readonly Array<string> includedFiles = list<string>(): new;

   public Line CurrentLine => SourceCode.CurrentLine;
   public RootNodeList Nodes => RootNode.RootNodes;

   public SourceCodeNode RootNode
   public ISourceCodeDocument SourceCode
   public IParserLogger Logger

   constructor(logger: IParserLogger, sourceCodeDocument: ISourceCodeDocument) {
     Logger = logger ?? throw new Error(nameof(logger));
     SourceCode = sourceCodeDocument ?? throw new Error(nameof(sourceCodeDocument));

     RootNode = new SourceCodeNode();
   }

   public addFileIncluded(fileName: string): void {
     let path = NormalizePath(fileName);

     includedFiles.Add(path);
   }

   public isFileIncluded(fileName: string): boolean {
     return includedFiles.contains(NormalizePath(fileName));
   }

   private static normalizePath(fileName: string): string {
     return Path.GetFullPath(fileName)
       .TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
   }
}

File: src/Lexy.Compiler/Parser/ParserContext.cs
using System;
using System.Collections.Generic;
using Lexy.Compiler.FunctionLibraries;
using Lexy.Compiler.Infrastructure;
using Lexy.Compiler.Language;

namespace Lexy.Compiler.Parser;

public class ParserContext : IParserContext
{
    private readonly IFileSystem fileSystem;

    private readonly IList<string> includedFiles = new List<string>();

    public ILibraries Libraries { get; }

    public ComponentNodeList Nodes => RootNode.ComponentNodes;
    public ILineFilter LineFilter { get; private set; }

    public LexyScriptNode RootNode { get; }
    public IParserLogger Logger { get; }
    public ParseOptions Options { get; }

    public ParserContext(IParserLogger logger, IFileSystem fileSystem, ILibraries libraries, ParseOptions options)
    {
        this.fileSystem = fileSystem ?? throw new ArgumentNullException(nameof(fileSystem));
        Logger = logger ?? throw new ArgumentNullException(nameof(logger));
        Libraries = libraries ?? throw new ArgumentNullException(nameof(libraries));

        Options = options ?? ParseOptions.Default();

        RootNode = new LexyScriptNode();
        LineFilter = new DefaultLineFilter();
    }

    public void AddFileIncluded(string fileName)
    {
        var path = NormalizePath(fileName);

        includedFiles.Add(path);
    }

    public bool IsFileIncluded(string fileName)
    {
        return includedFiles.Contains(NormalizePath(fileName));
    }

    private string NormalizePath(string fileName)
    {
        return fileSystem.GetFullPath(fileName);
    }

    public void SetFileLineFilter(string fileName)
    {
        LineFilter = fileName.EndsWith(LexySourceDocument.MarkdownExtension) ? new MarkdownLineFilter() : new DefaultLineFilter();
    }
}